### ISSUE TITLE
Auter --apply tries to install previous prepared rpms

### ISSUE THREAD

**Issue by reaperzn** [rackerlabs/auter#66]
If auter is set to only install from prep, after a prep has completed and an apply is run, if a second apply is attempted, in some cases, auter will fail:
```
[root@auter-test7 ~]# auter --apply
INFO: Running with: /usr/bin/auter --apply
INFO: Running in an interactive shell, disabling all random sleeps
INFO: Running Pre-Apply script /etc/auter/pre-apply.d/01-configsnap-pre
INFO: Applying updates
ERROR: Updates failed. Exiting.
```

In this case there was a long list of updates from a previous prep:
```
[root@auter-test7 ~]# ls -lah /var/cache/auter/default/ | wc -l
82
```
By setting the ```set -x``` option I found that this was the command that was being run:
```
[root@auter-test7 ~]# yum install -y --disableexcludes=all /var/cache/auter/default/bind-libs-9.9.4-50.el7_3.1.x86_64.rpm /var/cache/auter/default/bind-libs-lite-9.9.4-50.el7_3.1.x86_64.rpm /var/cache/auter/default/bind-license-9.9.4-50.el7_3.1.noarch.rpm /var/cache/auter/default/bind-utils-9.9.4-50.el7_3.1.x86_64.rpm /var/cache/auter/default/ca-certificates-2017.2.14-70.1.el7_3.noarch.rpm /var/cache/auter/default/chkconfig-1.7.2-1.el7_3.1.x86_64.rpm /var/cache/auter/default/configsnap-0.13-1.el7.noarch.rpm /var/cache/auter/default/device-mapper-1.02.135-1.el7_3.5.x86_64.rpm /var/cache/auter/default/device-mapper-event-1.02.135-1.el7_3.5.x86_64.rpm /var/cache/auter/default/device-mapper-event-libs-1.02.135-1.el7_3.5.x86_64.rpm /var/cache/auter/default/device-mapper-libs-1.02.135-1.el7_3.5.x86_64.rpm /var/cache/auter/default/dmidecode-3.0-2.1.el7_3.x86_64.rpm /var/cache/auter/default/dracut-033-463.el7_3.2.x86_64.rpm /var/cache/auter/default/dracut-config-rescue-033-463.el7_3.2.x86_64.rpm /var/cache/auter/default/dracut-network-033-463.el7_3.2.x86_64.rpm /var/cache/auter/default/epel-release-7-10.noarch.rpm /var/cache/auter/default/firefox-52.2.0-1.el7.centos.x86_64.rpm /var/cache/auter/default/firewalld-0.4.3.2-8.1.el7_3.3.noarch.rpm /var/cache/auter/default/firewalld-filesystem-0.4.3.2-8.1.el7_3.3.noarch.rpm /var/cache/auter/default/gawk-4.0.2-4.el7_3.1.x86_64.rpm /var/cache/auter/default/glibc-2.17-157.el7_3.5.x86_64.rpm /var/cache/auter/default/glibc-common-2.17-157.el7_3.5.x86_64.rpm /var/cache/auter/default/graphite2-1.3.10-1.el7_3.x86_64.rpm /var/cache/auter/default/grubby-8.28-21.el7_3.x86_64.rpm /var/cache/auter/default/gtk3-3.14.13-20.el7_3.1.x86_64.rpm /var/cache/auter/default/initscripts-9.49.37-1.el7_3.1.x86_64.rpm /var/cache/auter/default/irqbalance-1.0.7-6.el7_3.1.x86_64.rpm /var/cache/auter/default/jasper-libs-1.900.1-30.el7_3.x86_64.rpm /var/cache/auter/default/kernel-3.10.0-514.26.2.el7.x86_64.rpm /var/cache/auter/default/kernel-tools-3.10.0-514.26.2.el7.x86_64.rpm /var/cache/auter/default/kernel-tools-libs-3.10.0-514.26.2.el7.x86_64.rpm /var/cache/auter/default/kpartx-0.4.9-99.el7_3.3.x86_64.rpm /var/cache/auter/default/libblkid-2.23.2-33.el7_3.2.x86_64.rpm /var/cache/auter/default/libc-client-2007f-16.el7.x86_64.rpm /var/cache/auter/default/libgudev1-219-30.el7_3.9.x86_64.rpm /var/cache/auter/default/libmount-2.23.2-33.el7_3.2.x86_64.rpm /var/cache/auter/default/libnetfilter_conntrack-1.0.6-1.el7_3.x86_64.rpm /var/cache/auter/default/libtirpc-0.2.4-0.8.el7_3.x86_64.rpm /var/cache/auter/default/libtomcrypt-1.17-25.el7.x86_64.rpm /var/cache/auter/default/libtommath-0.42.0-5.el7.x86_64.rpm /var/cache/auter/default/libtool-ltdl-2.4.2-22.el7_3.x86_64.rpm /var/cache/auter/default/libuuid-2.23.2-33.el7_3.2.x86_64.rpm /var/cache/auter/default/lvm2-2.02.166-1.el7_3.5.x86_64.rpm /var/cache/auter/default/lvm2-libs-2.02.166-1.el7_3.5.x86_64.rpm /var/cache/auter/default/NetworkManager-1.4.0-20.el7_3.x86_64.rpm /var/cache/auter/default/NetworkManager-libnm-1.4.0-20.el7_3.x86_64.rpm /var/cache/auter/default/NetworkManager-team-1.4.0-20.el7_3.x86_64.rpm /var/cache/auter/default/NetworkManager-tui-1.4.0-20.el7_3.x86_64.rpm /var/cache/auter/default/nss-3.28.4-1.2.el7_3.x86_64.rpm /var/cache/auter/default/nss-sysinit-3.28.4-1.2.el7_3.x86_64.rpm /var/cache/auter/default/nss-tools-3.28.4-1.2.el7_3.x86_64.rpm /var/cache/auter/default/nss-util-3.28.4-1.0.el7_3.x86_64.rpm /var/cache/auter/default/ntpdate-4.2.6p5-25.el7.centos.2.x86_64.rpm /var/cache/auter/default/ntsysv-1.7.2-1.el7_3.1.x86_64.rpm /var/cache/auter/default/openssh-6.6.1p1-35.el7_3.x86_64.rpm /var/cache/auter/default/openssh-clients-6.6.1p1-35.el7_3.x86_64.rpm /var/cache/auter/default/openssh-server-6.6.1p1-35.el7_3.x86_64.rpm /var/cache/auter/default/php-PHPMailer-5.2.24-1.el7.noarch.rpm /var/cache/auter/default/polkit-0.112-12.el7_3.x86_64.rpm /var/cache/auter/default/pulseaudio-libs-6.0-9.el7_3.x86_64.rpm /var/cache/auter/default/python2-boto-2.45.0-3.el7.noarch.rpm /var/cache/auter/default/python-dmidecode-3.10.13-12.el7_3.x86_64.rpm /var/cache/auter/default/python-firewall-0.4.3.2-8.1.el7_3.3.noarch.rpm /var/cache/auter/default/python-perf-3.10.0-514.26.2.el7.x86_64.rpm /var/cache/auter/default/qemu-img-1.5.3-126.el7_3.10.x86_64.rpm /var/cache/auter/default/rdma-7.3_4.7_rc2-6.el7_3.noarch.rpm /var/cache/auter/default/rpcbind-0.2.0-38.el7_3.1.x86_64.rpm /var/cache/auter/default/selinux-policy-3.13.1-102.el7_3.16.noarch.rpm /var/cache/auter/default/selinux-policy-targeted-3.13.1-102.el7_3.16.noarch.rpm /var/cache/auter/default/sudo-1.8.6p7-23.el7_3.x86_64.rpm /var/cache/auter/default/systemd-219-30.el7_3.9.x86_64.rpm /var/cache/auter/default/systemd-libs-219-30.el7_3.9.x86_64.rpm /var/cache/auter/default/systemd-python-219-30.el7_3.9.x86_64.rpm /var/cache/auter/default/systemd-sysv-219-30.el7_3.9.x86_64.rpm /var/cache/auter/default/tuned-2.7.1-3.el7_3.2.noarch.rpm /var/cache/auter/default/tzdata-2017b-1.el7.noarch.rpm /var/cache/auter/default/util-linux-2.23.2-33.el7_3.2.x86_64.rpm /var/cache/auter/default/wordpress-4.8.1-1.el7.noarch.rpm /var/cache/auter/default/xfsprogs-4.5.0-10.el7_3.x86_64.rpm
```

The summary of that yum command is:
```
Resolving Dependencies
--> Running transaction check
---> Package kernel.x86_64 0:3.10.0-514.26.2.el7 will be installed
--> Finished Dependency Resolution

Dependencies Resolved

=======================================================================================================================
 Package          Arch             Version                          Repository                                    Size
=======================================================================================================================
Reinstalling:
 kernel           x86_64           3.10.0-514.26.2.el7              /kernel-3.10.0-514.26.2.el7.x86_64           148 M

Transaction Summary
=======================================================================================================================
Reinstall  1 Package

Total size: 148 M
Installed size: 148 M
Downloading packages:
Running transaction check
Running transaction test


Transaction check error:
  package kernel-3.10.0-514.26.2.el7.x86_64 is already installed

Error Summary
-------------

```
This is essentially because in the apply_updates function we have the following logic:
```
  if [[ "${ONLYINSTALLFROMPREP}" == "yes" ]]; then
    RC=0
    if [[ $(ls -1 ${DOWNLOADDIR}/${CONFIGSET}/*.rpm | wc -l) -gt 0 ]]; then
      RPMS="${DOWNLOADDIR}/${CONFIGSET}/*.rpm"
      RC=100
    fi 
    # When passing RPMs to dnf/yum, the update verb won't install any that aren't already
    # installed (i.e. dependencies of other packages). Instead we need to use install.
    UPDATEACTION="install"
  else
    local RC=$(${PACKAGE_MANAGER} check-update ${PACKAGEMANAGEROPTIONS} &>/dev/null; echo $?)
    UPDATEACTION="update"
  fi
```
We are statically setting the RC to 100 (return code of yum check-update when there are updates to be applied) if there are files in the auter rpm cache. There is no test to see if those packages require updating. We should look into adjusting the line to replicate the same actions taken if ONLYINSTALLFROMPREP is set to no:

proposed code:
```
  if [[ "${ONLYINSTALLFROMPREP}" == "yes" ]]; then
    RC=0
    if [[ $(ls -1 ${DOWNLOADDIR}/${CONFIGSET}/*.rpm | wc -l) -gt 0 ]]; then
      RPMS="${DOWNLOADDIR}/${CONFIGSET}/*.rpm"
      RC=$(${PACKAGE_MANAGER} check-update ${PACKAGEMANAGEROPTIONS} $(rpm -qp ${RPMS})&>/dev/null; echo $?)
    fi
    # When passing RPMs to dnf/yum, the update verb won't install any that aren't already
    # installed (i.e. dependencies of other packages). Instead we need to use install.
    UPDATEACTION="install"
  else
    local RC=$(${PACKAGE_MANAGER} check-update ${PACKAGEMANAGEROPTIONS} &>/dev/null; echo $?)
    UPDATEACTION="update"
  fi
```

With that change I get the following:
```
[root@auter-test7 ~]# auter --apply
INFO: Running with: /usr/bin/auter --apply
INFO: Running in an interactive shell, disabling all random sleeps
INFO: No updates are available to be applied.
```
However if I downgrade a single package in that list I still get the same problem:
```
INFO: Running with: /usr/bin/auter --apply
INFO: Running in an interactive shell, disabling all random sleeps
INFO: Running Pre-Apply script /etc/auter/pre-apply.d/01-configsnap-pre
INFO: Applying updates
ERROR: Updates failed. Exiting.
```
Manual retry using yum:
```
Dependencies Resolved

=======================================================================================================================
 Package            Arch             Version                        Repository                                    Size
=======================================================================================================================
Updating:
 xfsprogs           x86_64           4.5.0-10.el7_3                 /xfsprogs-4.5.0-10.el7_3.x86_64              3.9 M
Reinstalling:
 kernel             x86_64           3.10.0-514.26.2.el7            /kernel-3.10.0-514.26.2.el7.x86_64           148 M

Transaction Summary
=======================================================================================================================
Upgrade    1 Package
Reinstall  1 Package

Total size: 152 M
Downloading packages:
Running transaction check
Running transaction test


Transaction check error:
  package kernel-3.10.0-514.26.2.el7.x86_64 is already installed

```

Any ideas for a proper fix for this?

**Comment by reaperzn** [comment]
I think I may have found it but its not very pretty... Overwriting a variables:

```
  if [[ "${ONLYINSTALLFROMPREP}" == "yes" ]]; then
    RC=0
    if [[ $(ls -1 ${DOWNLOADDIR}/${CONFIGSET}/*.rpm | wc -l) -gt 0 ]]; then
      RPMS="${DOWNLOADDIR}/${CONFIGSET}/*.rpm"
      UPDATESPENDING=$(${PACKAGE_MANAGER} -q check-update ${PACKAGEMANAGEROPTIONS} $(rpm -qp ${RPMS})); RC=$?
      UPDATESPENDING=$(echo "$UPDATESPENDING" | awk '/^[[:alpha:]]/ {print $1"-"$2}')
      UPDATESPENDING=$(echo "$UPDATESPENDING" | sed 's/\.x86_64//g; s/\.noarch//g; s/\.[0-9]86//g' | xargs | tr " " "|")
      RPMS=$(ls -1 $RPMS | egrep "${UPDATESPENDING}" | xargs)
    fi
    # When passing RPMs to dnf/yum, the update verb won't install any that aren't already
    # installed (i.e. dependencies of other packages). Instead we need to use install.
    UPDATEACTION="install"
  else
    local RC=$(${PACKAGE_MANAGER} check-update ${PACKAGEMANAGEROPTIONS} &>/dev/null; echo $?)
    UPDATEACTION="update"
  fi
```
This seems to work on RHEL7, I have not tested on other distros yet.

**Comment by reaperzn** [comment]
I think I may have found it but its not very pretty... Overwriting a variables:

```
  if [[ "${ONLYINSTALLFROMPREP}" == "yes" ]]; then
    RC=0
    if [[ $(ls -1 ${DOWNLOADDIR}/${CONFIGSET}/*.rpm | wc -l) -gt 0 ]]; then
      RPMS="${DOWNLOADDIR}/${CONFIGSET}/*.rpm"
      UPDATESPENDING=$(${PACKAGE_MANAGER} -q check-update ${PACKAGEMANAGEROPTIONS} $(rpm -qp ${RPMS})); RC=$?
      UPDATESPENDING=$(echo "$UPDATESPENDING" | awk '/^[[:alpha:]]/ {print $1"-"$2}')
      UPDATESPENDING=$(echo "$UPDATESPENDING" | sed 's/\.x86_64//g; s/\.noarch//g; s/\.[0-9]86//g' | xargs | tr " " "|")
      RPMS=$(ls -1 $RPMS | egrep "${UPDATESPENDING}" | xargs)
    fi
    # When passing RPMs to dnf/yum, the update verb won't install any that aren't already
    # installed (i.e. dependencies of other packages). Instead we need to use install.
    UPDATEACTION="install"
  else
    local RC=$(${PACKAGE_MANAGER} check-update ${PACKAGEMANAGEROPTIONS} &>/dev/null; echo $?)
    UPDATEACTION="update"
  fi
```
This seems to work on RHEL7, I have not tested on other distros yet.

**Comment by reaperzn** [comment]
I just re-looked at this code and this wont work...  this does a check-update which hit the same issue as the comment directly after... the update or in this case check-update verb won't install any that aren't already installed.

**Comment by reaperzn** [comment]
Just so I dont forget, I had a discussion with @piersc and we were considering doing a manual check for the kernel version and if the rpm version is already installed then remove that package from the transaction.

**Comment by reaperzn** [comment]
finally should be resolved in #99
